### ISSUE TITLE
Configuration from the environment?

### ISSUE THREAD

**Issue by kevineaton** [Leantime/leantime#199]
Would you be interested in accepting parameters from the environment for the configuration, with some sane defaults? It would make the docker image much easier to use. For example, the docker image start.sh file appears to physically rewrite the file:
https://github.com/Leantime/docker-leantime/blob/master/start.sh

`sed -i 's/$dbHost="localhost"/$dbHost="'"${DB_HOST}"'"/g' config/configuration.php && \` 

for example. This means that to use the docker image to any real extent, we need to manually mount a file since only a few of the options are exposed.

 Instead, if the configuration.php file checked the environment first then fell back to the current defaults, that would make it easier I believe. (I am referring to https://github.com/Leantime/leantime/blob/master/config/configuration.sample.php)

I haven't written PHP in nearly a decade, but I could take a stab at it if you don't have time. I do not know the performance ramifications, as I typically work in Go and have a singleton configuration object. However, I would essentially create a class helper method to look at the environment and then add a constructor for all of the values, something like:

```php
        function __construct(){
            $this->sitename = $this->configEnvironmentHelper("LEAN_SITENAME", $this->sitename);
        }

        private function configEnvironmentHelper($envVar, $default) {
            return isset($_ENV[$envVar]) ? $_ENV[$envVar] : $default;
        }

```

What are your thoughts?

**Comment by marcelfolaron** [comment]
I think this is a great idea. Wondering if it would be easier to just keep a docker specific config.sample that uses the environment variables. Start.sh would then just rename that file instead of the regular sample config. @jjensen90 your thoughts?

**Comment by jjensen90** [comment]
This is a good idea, and was already on my list for moving Leantime towards the 12 factor app principles. Runtime configuration injection via env is definitely on my roadmap. 

Before implementing a home-grown env var config I would consider using the `.env` file format for config, similar to Laravel, and letting the app load env vars from that, which are then overridden by any env vars set manually (in docker compose, cli. K8s manifest, etc). There is a popular package for this by symfony, as well as some other open source options. If you're interested we could work together on an implementation PoC.

**Comment by kevineaton** [comment]
I am going to be honest, my PHP experience is fairly dated. The last time I worked in PHP, it was PHP 5, and Symfony and crew were only just starting to become popular. I have been experimenting with a a simple implementation and honestly, getting the variables to appear in the getenv() calls has been a challenge; I am running it behind NGINX with php-fpm on Ubuntu. It looks like php-fpm doesn't always pass over the values, so users would have to modify their php-fpm config file, which may not be great in userland. Even then, it seems inconsistent, although this likely wouldn't be an issue in the Docker repo due to the installation of Apache2, which I haven't touched in awhile as well.

So, I am far from the expert in this area and I am not up to date with best practices in PHP. I defer to your wisdom. We are experimenting with Leantime at my organization, and haven't used the docker image because we also want S3 and SMTP. So we are running the app as it stands. I can poke around and try to learn, but I am sure you guys know the PHP ecosystem far better than I!

**Comment by gmkumar2005** [comment]
Hi all I am able to make it work with below changes

1) File --> /etc/php7/php-fpm.d/www.conf should have `clear_env = no`
2) Need to use constructor in  the configuration.php 


``` php
<?php
namespace leantime\core;

class config
{
public $dbHost = null; 
public function __construct() {
        $this->dbHost = getenv('DATABASE_HOST');
    }
}
```

I have tested with docker php:7.3-fpm-alpine on openshift

**Comment by kevineaton** [comment]
Here is where I ended up:
https://github.com/kevineaton/leantime/blob/master/config/configuration.sample.php

I opened a draft PR for feedback:
https://github.com/Leantime/leantime/pull/201

**Comment by gmkumar2005** [comment]
> Here is where I ended up:
> https://github.com/kevineaton/leantime/blob/master/config/configuration.sample.php
> 
> I opened a draft PR for feedback:
> #201

I have verified the fix on Leantime-V2.0.15 release on openshift. It works.

**Comment by kevineaton** [comment]
With this being merged in https://github.com/Leantime/leantime/pull/201 we can probably close this.
